java-generator/core/src/main/java/io/fabric8/java/generator/FileJavaGenerator.java

```diff
@@ -83,32 +85,50 @@ private void runOnSingleSource(File source, File basePath) {
         resources.add(deserialized);
       }
 
-      resources.parallelStream()
-          .forEach(
-              rawResource -> {
-                if (rawResource != null && rawResource instanceof HasMetadata) {
-                  final HasMetadata resource = (HasMetadata) rawResource;
+      writeCRCompilationUnits(basePath, generateWritableCRCompilationUnits(resources));
 
-                  if (resource != null && resource.getKind()
-                      .toLowerCase(Locale.ROOT)
-                      .equals("customresourcedefinition")) {
-                    CustomResourceDefinition crd = (CustomResourceDefinition) resource;
-
-                    final String basePackage = groupToPackage(crd.getSpec().getGroup());
-
-                    crGeneratorRunner.generate(crd, basePackage).parallelStream()
-                        .forEach(w -> w.writeAllJavaClasses(basePath));
-                  } else {
-                    LOGGER.warn("Not generating nothing for resource of kind: {}", resource.getKind());
-                  }
-                } else {
-                  LOGGER.warn("Not generating nothing for unrecognized resource: {}", Serialization.asYaml(rawResource));
-                }
-              });
     } catch (FileNotFoundException e) {
       throw new JavaGeneratorException("File " + source.getAbsolutePath() + " not found", e);
     } catch (IOException e) {
       throw new JavaGeneratorException("Exception reading " + source.getAbsolutePath(), e);
     }
   }
+
+  private List<WritableCRCompilationUnit> generateWritableCRCompilationUnits(List<Object> resources) {
+    return resources.parallelStream()
+        .flatMap(
+            rawResource -> {
+              if (rawResource != null && rawResource instanceof HasMetadata) {
+                final HasMetadata resource = (HasMetadata) rawResource;
+
+                if (resource != null && resource.getKind()
+                    .toLowerCase(Locale.ROOT)
+                    .equals("customresourcedefinition")) {
+                  CustomResourceDefinition crd = (CustomResourceDefinition) resource;
+
+                  final String basePackage = groupToPackage(crd.getSpec().getGroup());
+                  return crGeneratorRunner.generate(crd, basePackage).stream();
+                } else {
+                  LOGGER.warn("Not generating nothing for resource of kind: {}", resource.getKind());
+                }
+              } else {
+                LOGGER.warn("Not generating nothing for unrecognized resource: {}", Serialization.asYaml(rawResource));
+              }
+              return Stream.empty();
+            })
+      .collect(Collectors.toList());
+  }
+
+  private void writeCRCompilationUnits(File basePath, List<WritableCRCompilationUnit> wCUs) {
+    CompletableFuture<Void>[] futures = new CompletableFuture[wCUs.size()];
+
+    int i = 0;
```

Comment:
[nitpick] can you, please, move `wCUs.size()` to a val (e.g. `total`) and reuse it as a boundary in the following `for` loops?

Reply:
Done.